src/components/menu/measureNumberFormat/MeasureNumberFormat.tsx

```diff
@@ -0,0 +1,184 @@
+import { ISeparators } from "@gooddata/numberjs";
+
+import { IFormatPreset, IToggleButtonProps, PresetType } from "./typings";
+import { PresetsDropdown } from "./presetsDropdown/PresetsDropdown";
+import { IntlWrapper } from "../../core/base/IntlWrapper";
+import { CustomFormatDialog } from "./customFormatDialog/CustomFormatDialog";
+import { IPositioning } from "../../../typings/positioning";
+import { IFormatTemplate } from "./customFormatDialog/formatTemplatesDropdown/FormatTemplatesDropdown";
+
+const CUSTOM_FORMAT_PRESET_LOCAL_IDENTIFIER = "customFormat";
+
+export interface IMeasureNumberFormatOwnProps {
+    toggleButton: React.ComponentType<IToggleButtonProps>;
+    presets: ReadonlyArray<IFormatPreset>;
+    separators: ISeparators;
+    selectedFormat: string;
+    setFormat: (format: string) => void;
+    anchorElementSelector?: string;
+    presetsDropdownPositioning?: IPositioning[];
+    customFormatDialogPositioning?: IPositioning[];
+    defaultCustomFormat?: string;
+    documentationLink?: string;
+    templates?: ReadonlyArray<IFormatTemplate>;
+    locale?: string;
+}
+
+export type MeasureNumberFormatProps = IMeasureNumberFormatOwnProps & WrappedComponentProps;
+
+interface IMeasureNumberFormatState {
+    showDropdown: boolean;
+    showCustomFormatDialog: boolean;
+    selectedPreset: IFormatPreset;
+}
+
+class WrappedMeasureNumberFormat extends React.PureComponent<
+    MeasureNumberFormatProps,
+    IMeasureNumberFormatState
+> {
+    private toggleButtonEl: EventTarget;
+
+    constructor(props: MeasureNumberFormatProps) {
+        super(props);
+
+        this.state = {
+            selectedPreset: this.findSelectedPreset(),
+            showDropdown: false,
+            showCustomFormatDialog: false,
+        };
+    }
+
+    public render() {
+        const {
+            toggleButton: ToggleButton,
+            anchorElementSelector,
+            presets,
+            separators,
+            selectedFormat,
+            defaultCustomFormat,
+            presetsDropdownPositioning,
+            customFormatDialogPositioning,
+            documentationLink,
+            templates,
+            intl,
+        } = this.props;
+        const { showDropdown, showCustomFormatDialog, selectedPreset } = this.state;
+        const buttonText = intl.formatMessage(
+            { id: "measureNumberFormat.buttonLabel" },
+            { selectedFormatPresetName: selectedPreset.name },
+        );
+        const anchorEl = anchorElementSelector || this.toggleButtonEl;
+        return (
+            <>
+                <ToggleButton
+                    text={buttonText}
+                    isOpened={showDropdown || showCustomFormatDialog}
+                    toggleDropdown={this.toggleDropdownOpened}
+                />
+                {showDropdown && (
+                    <PresetsDropdown
+                        presets={[...presets, this.getCustomFormatPreset()]}
```

Comment:
This producers new array each time functional component is called. We decided to not do memoization if we don't do any measurement of performance to have more readable code?

Reply:
Changed the way presets are rendered.